Security key doesn't show up when allowCredentials transports contain "Internal"

pascoej opened the pull request:
#### 442e5675cbb3f411a08db3c8bc4b9ed0afdfb95a
<pre>
Security key doesn&apos;t show up when allowCredentials transports contain &quot;Internal&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=271805">https://bugs.webkit.org/show_bug.cgi?id=271805</a>
<a href="https://rdar.apple.com/124265507">rdar://124265507</a>

Reviewed by Brent Fulgham.

This condition was written wrong. We should submit a cross-platform request if transports are empty, or if
they contain any transport that is not &quot;Internal&quot; or &quot;Hybrid,&quot; which are both considered platform requests
by system API.

* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::WebAuthenticatorCoordinatorProxy::requestsForAssertion):

Canonical link: <a href="https://commits.webkit.org/276894@main">https://commits.webkit.org/276894@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6934e2aaacd131dd4d99617051afef5075c78e84

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46068 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25198 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48653 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48739 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42108 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29517 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22638 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37671 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46646 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22290 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39748 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18861 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19689 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4112 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42388 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41195 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50530 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21064 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17583 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/44849 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22364 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43749 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10206 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22723 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22058 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -345,7 +345,7 @@ static inline ASAuthorizationPublicKeyCredentialLargeBlobSupportRequirement toAS
         if (credential.transports.contains(AuthenticatorTransport::Internal) || credential.transports.isEmpty())
             [platformAllowedCredentials addObject:adoptNS([allocASAuthorizationPlatformPublicKeyCredentialDescriptorInstance() initWithCredentialID:toNSData(credential.id).get()]).get()];
-        if (credential.transports.isEmpty() || !credential.transports.contains(AuthenticatorTransport::Internal)) {
+        if (credential.transports.isEmpty() || credential.transports.containsIf([](auto transport) { return transport != AuthenticatorTransport::Internal && transport != AuthenticatorTransport::Hybrid; })) {
```
I wonder if this new logic (neither `Internal` not `Hybrid`) could be represented as `isNotPlatformRequest(credential.transports)`?

That might be slightly easier to understand when reading, but it's just a suggestion.

brentfulgham reviewed (approved):
Looks good! I had a suggestion to perhaps make the logic a little more readable, but that's just a suggestion. r=me

pascoej committed:
Security key doesn't show up when allowCredentials transports contain "Internal"
https://bugs.webkit.org/show_bug.cgi?id=271805
rdar://124265507

Reviewed by Brent Fulgham.

This condition was written wrong. We should submit a cross-platform request if transports are empty, or if
they contain any transport that is not "Internal" or "Hybrid," which are both considered platform requests
by system API.

* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
(WebKit::WebAuthenticatorCoordinatorProxy::requestsForAssertion):

Canonical link: https://commits.webkit.org/276894@main
